File: macos/AgentMeter/Presentation/SetupFlow.swift
import Foundation
import CoreFoundation
import Observation

enum SetupCompletion {
  static let key = "setupCompleted"
  static func isComplete(_ defaults: UserDefaults) -> Bool {
    guard let value = defaults.object(forKey: key) as? NSNumber,
      CFGetTypeID(value) == CFBooleanGetTypeID() else { return false }
    return value.boolValue
  }
  static func hasPreferences(_ values: [String: Any]) -> Bool {
    if let value = values["appearance"] as? String, AppAppearance(rawValue: value) != nil { return true }
    return ["notchEnabled", "menuEnabled"].contains { key in
      guard let value = values[key] as? NSNumber else { return false }
      return CFGetTypeID(value) == CFBooleanGetTypeID()
    }
  }
  static func recognizeExisting(defaults: UserDefaults, old: [String: Any]) {
    guard defaults.object(forKey: key) == nil else { return }
    let current = Dictionary(uniqueKeysWithValues: ["appearance", "notchEnabled", "menuEnabled"].compactMap {
      key in defaults.object(forKey: key).map { (key, $0) }
    })
    // Persist false for a fresh install too: changing preferences partway through setup
    // must not make an incomplete first run look like an existing installation later.
    defaults.set(hasPreferences(current) || hasPreferences(old), forKey: key)
  }
}

enum SetupStep: Equatable {
  case welcome, providers, codex, claude, verify, preferences, done
}

enum SetupStatus: String {
  case checking = "Checking…"
  case notInstalled = "Not installed"
  case signedOut = "Installed — sign in required"
  case ready = "Ready"
  case unavailable = "Unable to verify — check again"
  init(snapshot: UsageSnapshot) {
    switch snapshot.state {
    case .live: self = snapshot.reading == nil ? .unavailable : .ready
    case .notInstalled: self = .notInstalled
    case .signedOut: self = .signedOut
    case .loading: self = .checking
    case .stale, .unavailable: self = .unavailable
    }
  }
}

@MainActor @Observable final class SetupFlow {
  private let defaults: UserDefaults
  private(set) var step: SetupStep = .welcome
  var selected: Set<ProviderID> = Set(ProviderID.allCases)
  var needsAutomaticSetup: Bool { !SetupCompletion.isComplete(defaults) }
  var steps: [SetupStep] {
    [.welcome, .providers] + (selected.contains(.codex) ? [.codex] : [])
      + (selected.contains(.claude) ? [.claude] : []) + [.verify, .preferences, .done]
  }
  init(defaults: UserDefaults = .standard) { self.defaults = defaults }
  func reopen() { step = .welcome }
  func next() {
    guard let index = steps.firstIndex(of: step), index + 1 < steps.count else { return }
    step = steps[index + 1]
  }
  func back() {
    guard let index = steps.firstIndex(of: step), index > 0 else { return }
    step = steps[index - 1]
  }
  func complete() { defaults.set(true, forKey: SetupCompletion.key) }
}

// Copy-only instructions verified against official docs on 2026-09-23.
enum ProviderSetup {
  static func install(_ provider: ProviderID) -> String {
    provider == .codex ? "brew install --cask codex" : "curl -fsSL https://claude.ai/install.sh | bash"
  }
  static func login(_ provider: ProviderID) -> String {
    provider == .codex ? "codex login" : "claude auth login"
  }
  static func documentation(_ provider: ProviderID) -> URL {
    URL(string: provider == .codex ? "https://learn.chatgpt.com/docs/codex/cli" : "https://code.claude.com/docs/en/quickstart")!
  }
}

// Deliberately no Reading, Installation, raw errors, or log text in this export.
struct SetupDiagnostic {
  let detected: String
  let authentication: String
  let usage: String
  let failure: String
  let result: String
  init(_ snapshot: UsageSnapshot) {
    switch snapshot.state {
    case .live where snapshot.reading != nil:
      detected = "yes"; authentication = "verified"; usage = "available"; result = "success"
    case .notInstalled:
      detected = "no"; authentication = "unknown"; usage = "unavailable"; result = "failure"
    case .signedOut:
      detected = "yes"; authentication = "signed-out"; usage = "unavailable"; result = "failure"
    case .stale:
      detected = "unknown"; authentication = "unknown"; usage = "stale"; result = "failure"
    case .loading:
      detected = "unknown"; authentication = "unknown"; usage = "checking"; result = "pending"
    default:
      detected = "unknown"; authentication = "unknown"; usage = "unavailable"; result = "failure"
    }
    failure = snapshot.failure?.rawValue ?? "none"
  }
  var summary: String { "Detected: \(detected) · Authentication: \(authentication) · Usage: \(usage)" }
}
enum SetupDiagnostics {
  static func numeric(_ value: String?) -> String {
    guard let value, value.count <= 32, !value.isEmpty,
      value.split(separator: ".", omittingEmptySubsequences: false).allSatisfy({
        !$0.isEmpty && $0.utf8.allSatisfy { (48...57).contains($0) }
      }) else { return "unknown" }
    return value
  }
  static func report(_ usage: [ProviderID: UsageSnapshot], version: String?, build: String?) -> String {
    let os = ProcessInfo.processInfo.operatingSystemVersion
    #if arch(arm64)
    let architecture = "arm64"
    #elseif arch(x86_64)
    let architecture = "x86_64"
    #else
    let architecture = "other"
    #endif
    var lines = ["Llumi diagnostics schema: 1", "App version: \(numeric(version))",
      "App build: \(numeric(build))", "OS: macOS \(os.majorVersion).\(os.minorVersion).\(os.patchVersion)",
      "Architecture: \(architecture)"]
    for provider in ProviderID.allCases {
      let value = SetupDiagnostic(usage[provider] ?? UsageSnapshot(provider: provider))
      lines += ["[\(provider.rawValue)]", "Detected: \(value.detected)",
        "Authentication: \(value.authentication)", "Usage: \(value.usage)",
        "Failure: \(value.failure)", "Last result: \(value.result)"]
    }
    return lines.joined(separator: "\n")
  }
}

File: windows/src/AgentMeter/SetupFlow.cs
using System.Text.Json;
using AgentMeter.Core;

namespace AgentMeter;

internal enum SetupStep { Welcome, Providers, Codex, Claude, Verify, Preferences, Done }
internal sealed class SetupFlow(SetupCompletionStore store)
{
    internal bool Codex { get; set; } = true;
    internal bool Claude { get; set; } = true;
    internal SetupStep Step { get; private set; } = SetupStep.Welcome;
    internal SetupStep[] Steps => new[] { SetupStep.Welcome, SetupStep.Providers }
        .Concat(Codex ? new[] { SetupStep.Codex } : []).Concat(Claude ? new[] { SetupStep.Claude } : [])
        .Concat([SetupStep.Verify, SetupStep.Preferences, SetupStep.Done]).ToArray();
    internal void Next() { var i = Array.IndexOf(Steps, Step); if (i >= 0 && i + 1 < Steps.Length) Step = Steps[i + 1]; }
    internal void Back() { var i = Array.IndexOf(Steps, Step); if (i > 0) Step = Steps[i - 1]; }
    internal void Reopen() => Step = SetupStep.Welcome;
    internal bool Complete() => store.Save(true);
}
internal sealed class SetupCompletionStore(string path)
{
    internal static SetupCompletionStore Default() => new(Path.Combine(PackagedEnvironment.DataDirectory, "setup-completed.json"));
    internal bool RecognizeExisting(bool existingPreferences)
    {
        if (File.Exists(path)) return IsComplete();
        Save(existingPreferences); // Persist fresh false before the user changes preferences.
        return existingPreferences;
    }
    internal bool IsComplete()
    {
        try
        {
            using var input = File.OpenRead(path);
            if (input.Length > 32) return false;
            using var value = JsonDocument.Parse(input);
            return value.RootElement.ValueKind == JsonValueKind.True;
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException or JsonException) { return false; }
    }
    internal bool Save(bool completed)
    {
        var temporary = path + ".tmp";
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(path)!);
            File.WriteAllText(temporary, completed ? "true" : "false");
            File.Move(temporary, path, true); return true;
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException) { return false; }
        finally { try { File.Delete(temporary); } catch (IOException) { } catch (UnauthorizedAccessException) { } }
    }
}

internal sealed record SetupDiagnostic(string Detected, string Authentication, string Usage, string Failure, string Result)
{
    internal static SetupDiagnostic From(ProviderState state)
    {
        var failure = state.Failure switch {
            FailureKind.None => "none", FailureKind.NotInstalled => "notInstalled", FailureKind.LoggedOut => "signedOut",
            FailureKind.Unsupported => "incompatible", FailureKind.Timeout => "timeout", FailureKind.Malformed => "malformed",
            FailureKind.ProcessExited => "processExited", _ => "unavailable" };
        if (state.Status == ProviderStatus.Ready && state.Snapshot is not null && !state.IsStale(DateTimeOffset.UtcNow))
            return new("yes", "verified", "available", failure, "success");
        if (state.Failure == FailureKind.NotInstalled) return new("no", "unknown", "unavailable", failure, "failure");
        if (state.Failure == FailureKind.LoggedOut) return new("yes", "signed-out", "unavailable", failure, "failure");
        if (state.Status == ProviderStatus.Loading) return new("unknown", "unknown", "checking", failure, "pending");
        return new("unknown", "unknown", state.Snapshot is null ? "unavailable" : "stale", failure, "failure");
    }
    internal string Status => Usage == "available" ? "Ready" : Detected == "no" ? "Not installed" :
        Authentication == "signed-out" ? "Installed — sign in required" : Usage == "checking" ? "Checking…" : "Unavailable";
    internal string Summary => $"{Status}\nDetected: {Detected} · Authentication: {Authentication} · Usage: {Usage}";
}
internal static class SetupDiagnostics
{
    internal static string Numeric(string? value) => value is { Length: > 0 and <= 32 } &&
        value.Split('.').All(p => p.Length > 0 && p.All(c => c is >= '0' and <= '9')) ? value : "unknown";
    internal static string Report(IReadOnlyList<ProviderState> states, string? version, string? build)
    {
        var lines = new List<string> { "Llumi diagnostics schema: 1", $"App version: {Numeric(version)}",
            $"App build: {Numeric(build)}", $"OS: Windows {Environment.OSVersion.Version}",
            $"Architecture: {System.Runtime.InteropServices.RuntimeInformation.ProcessArchitecture.ToString().ToLowerInvariant()}" };
        foreach (var provider in new[] { "Codex", "Claude" })
        {
            var state = states.FirstOrDefault(s => s.Name == provider || (provider == "Claude" && s.Name == "Claude Code"))
                ?? new ProviderState(provider, ProviderStatus.Loading);
            var value = SetupDiagnostic.From(state);
            lines.AddRange([$"[{provider.ToLowerInvariant()}]", $"Detected: {value.Detected}",
                $"Authentication: {value.Authentication}", $"Usage: {value.Usage}",
                $"Failure: {value.Failure}", $"Last result: {value.Result}"]);
        }
        return string.Join("\n", lines);
    }
}
